### ISSUE TITLE
[tablets] unlike repair-based replace operation, tablet-rebuild for replace streams data from all nodes, not just the node's datacenter

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylladb#18514]
As seen, for example, in https://jenkins.scylladb.com/job/scylla-master/job/tablets/job/gating-dtest-release-with-tablets/51/artifact/logs-full.release.002/1714699154601_repair_additional_test.py%3A%3ATestRepairAdditional%3A%3Atest_repair_streams_data_from_closest_node/node5.log
```
INFO  2024-05-03 01:19:09,934 [shard 0:strm] range_streamer - Tablet rebuild with [127.0.25.3, 127.0.25.1, 127.0.25.2] for keyspace=ks started, nodes_to_stream=3
```

Tablet rebuild does not limit the source replicas for rebuild to the replaced node's datacenter, as does repair-based replace.
(This is what this dtest is testing).
This isn't wrong, correctness-wise. It is even safer, but it is more expensive in inter-dc communication.

@tgrabiec is this by design?

**Comment by bhalevy** [comment]
> > @tgrabiec is this by design?
> 
> It was implemented like this for simplicity, we could restore vnodes logic if we think it's safe.

It's not safe in all cases - I actually opened https://github.com/scylladb/scylladb/issues/16826 to ask for a safer solution if the number of remaining replicas in the datacenter do not guarantee quorum, so we can do that with tablets ans adapt the dtest around that, testing both cases when tablets are enabled.

**Comment by mykaul** [comment]
@bhalevy - what's the status of this item? Who should own it?

**Comment by mykaul** [comment]
> [@bhalevy](https://github.com/bhalevy) - what's the status of this item? Who should own it?

ping @bhalevy

**Comment by bhalevy** [comment]
@Deexie / @denesb isn't that fixed?

**Comment by Deexie** [comment]
> [@Deexie](https://github.com/Deexie) / [@denesb](https://github.com/denesb) isn't that fixed?

Yes, it is. Currently tablet rebuild streams from the closest replica

**Comment by denesb** [comment]
@Deexie please link here the commit of the merged PR which fixes this, then close this issue as completed.

**Comment by Deexie** [comment]
https://github.com/scylladb/scylladb/pull/23187

**Comment by piodul** [comment]
Temporarily reopening to prevent more CI failures, unfortunately closing this issue un-skipped two tests that were skipped but not prepared to run with the rf-rack-valid restriction. See the failure here: https://jenkins.scylladb.com/job/scylla-2025.4/job/next/7/ (but also happens on master).

Failing dtests are test_repair_streams_data_from_all_nodes_with_tablets and test_repair_streams_data_from_closest_node.

**Comment by piodul** [comment]
> https://github.com/scylladb/scylladb/issues/18514#issuecomment-3382300415

FYI @denesb @Deexie

**Comment by Deexie** [comment]
Looking at https://github.com/scylladb/scylladb/issues/26459 I realized I might have misunderstood this issue.

Closing it, I assumed that "streaming" is another data exchanging method alongside repair (not a name of the exchanging itself - with repair or well... streaming). With that assumption the issue is completed - for tablet rebuild we stream from the closest node. But apart from streaming, we repair the tablet on all replicas.

For vnodes we may repair only the local DC (under some conditions). Is it what we expect from this issue? @bhalevy @denesb

**Comment by bhalevy** [comment]
We can close this issue once the test is fixed.

The gist of it was that there was huge data amplifications where the rebuilt node would receive data for **all other** replicas and would run out of space if compaction couldn't keep up with the rebuild rate.

**Comment by bhalevy** [comment]
Closing now that https://github.com/scylladb/scylladb/issues/26459 is fixed.
